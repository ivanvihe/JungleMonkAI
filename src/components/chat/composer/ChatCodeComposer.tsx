import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { estimateTokenCount } from '../../../utils/tokens';
import { copyToClipboard } from '../../../utils/clipboard';

interface ChatCodeComposerProps {
  value: string;
  placeholder?: string;
  appendToDraft: (value: string) => void;
  setDraft: (value: string) => void;
  onSend?: () => void;
  disabled?: boolean;
}

interface ToolbarAction {
  id: string;
  icon: string;
  label: string;
  shortcut?: string;
  onAction: () => void;
}

const CODE_BLOCK_TEMPLATE = '```\n\n```';

export const ChatCodeComposer: React.FC<ChatCodeComposerProps> = ({
  value,
  placeholder,
  appendToDraft,
  setDraft,
  onSend,
  disabled,
}) => {
  const textareaRef = useRef<HTMLTextAreaElement | null>(null);
  const [lastCopied, setLastCopied] = useState<number | null>(null);

  const tokenCount = useMemo(() => estimateTokenCount(value), [value]);

  const wrapSelectionWithCodeBlock = useCallback(
    (language?: string) => {
      const textarea = textareaRef.current;
      if (!textarea) {
        return;
      }
      const { selectionStart, selectionEnd } = textarea;
      const currentValue = textarea.value;
      const selectedText = currentValue.slice(selectionStart, selectionEnd);
      const normalizedLanguage = language?.trim() ?? '';
      const fenceHeader = `\`\`\`${normalizedLanguage}\n`;
      const fenceFooter = '\n```\n';
      const before = currentValue.slice(0, selectionStart);
      const after = currentValue.slice(selectionEnd);
      const needsLeadingNewline = before.length > 0 && !before.endsWith('\n');
      const headerPrefix = needsLeadingNewline ? `\n${fenceHeader}` : fenceHeader;
      const newValue = `${before}${headerPrefix}${selectedText}${fenceFooter}${after}`;
      setDraft(newValue);

      window.requestAnimationFrame(() => {
        const caretPosition = selectionStart + headerPrefix.length + selectedText.length;
        textarea.selectionStart = caretPosition;
        textarea.selectionEnd = caretPosition;
        textarea.focus();
      });
    },
    [setDraft],
  );

  const insertCodeBlockTemplate = useCallback(() => {
    const textarea = textareaRef.current;
    if (!textarea) {
      appendToDraft(`\n${CODE_BLOCK_TEMPLATE}`);
      return;
    }
    const { selectionStart, selectionEnd } = textarea;
    if (selectionStart !== selectionEnd) {
      wrapSelectionWithCodeBlock();
      return;
    }
    const currentValue = textarea.value;
    const before = currentValue.slice(0, selectionStart);
    const after = currentValue.slice(selectionEnd);
    const needsLeadingNewline = before.length > 0 && !before.endsWith('\n');
    const insertion = `${needsLeadingNewline ? '\n' : ''}\`\`\`\n\n\`\`\``;
    const newValue = `${before}${insertion}${after}`;
    setDraft(newValue);

    window.requestAnimationFrame(() => {
      const caretPosition = selectionStart + (needsLeadingNewline ? 5 : 4);
      textarea.selectionStart = caretPosition;
      textarea.selectionEnd = caretPosition;
      textarea.focus();
    });
  }, [appendToDraft, setDraft, wrapSelectionWithCodeBlock]);

  const handleKeyboardShortcuts = useCallback(
    (event: React.KeyboardEvent<HTMLTextAreaElement>) => {
      if (event.key === 'Enter' && (event.metaKey || event.ctrlKey)) {
        event.preventDefault();
        onSend?.();
        return;
      }

      if (event.key.toLowerCase() === 'c' && event.shiftKey && (event.metaKey || event.ctrlKey)) {
        event.preventDefault();
        wrapSelectionWithCodeBlock();
        return;
      }

      if (event.key === '`' && event.shiftKey && (event.metaKey || event.ctrlKey)) {
        event.preventDefault();
        insertCodeBlockTemplate();
      }
    },
    [insertCodeBlockTemplate, onSend, wrapSelectionWithCodeBlock],
  );

  const handleCopyAll = useCallback(() => {
    copyToClipboard(value)
      .then(() => {
        setLastCopied(Date.now());
      })
      .catch(() => {
        setLastCopied(null);
      });
  }, [value]);

  useEffect(() => {
    if (lastCopied === null) {
      return;
    }
    const timeout = window.setTimeout(() => setLastCopied(null), 2000);
    return () => window.clearTimeout(timeout);
  }, [lastCopied]);

  const toolbarActions: ToolbarAction[] = useMemo(
    () => [
      {
        id: 'code-block',
        icon: 'ðŸ§±',
        label: 'Bloque de cÃ³digo',
        shortcut: 'Ctrl+Shift+`',
        onAction: insertCodeBlockTemplate,
      },
      {
        id: 'wrap-selection',
        icon: 'ðŸª„',
        label: 'Envolver selecciÃ³n',
        shortcut: 'Ctrl+Shift+C',
        onAction: () => wrapSelectionWithCodeBlock(),
      },
      {
        id: 'copy',
        icon: 'ðŸ“‹',
        label: lastCopied ? 'Copiado!' : 'Copiar todo',
        onAction: handleCopyAll,
      },
    ],
    [handleCopyAll, insertCodeBlockTemplate, lastCopied, wrapSelectionWithCodeBlock],
  );

  return (
    <div className="chat-code-composer">
      <div className="chat-code-toolbar" role="toolbar" aria-label="Acciones del compositor">
        {toolbarActions.map(action => (
          <button
            key={action.id}
            type="button"
            className={`composer-toolbar-button${action.id === 'copy' && lastCopied ? ' is-active' : ''}`}
            onClick={action.onAction}
            disabled={disabled}
            title={action.shortcut ? `${action.label} (${action.shortcut})` : action.label}
            aria-label={action.label}
          >
            <span aria-hidden="true">{action.icon}</span>
          </button>
        ))}
        <div className="chat-code-toolbar-spacer" />
        <span className="chat-code-token-count" aria-live="polite">
          {tokenCount} tokens estimados
        </span>
      </div>
      <textarea
        ref={textareaRef}
        value={value}
        onChange={event => setDraft(event.target.value)}
        onKeyDown={handleKeyboardShortcuts}
        placeholder={placeholder}
        className="chat-code-textarea"
        rows={4}
        disabled={disabled}
      />
    </div>
  );
};

export default ChatCodeComposer;
