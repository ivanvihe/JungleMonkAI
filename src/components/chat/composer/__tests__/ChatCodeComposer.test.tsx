import React from 'react';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { ChatCodeComposer } from '../ChatCodeComposer';

vi.mock('../../../../utils/clipboard', () => ({
  copyToClipboard: vi.fn().mockResolvedValue(undefined),
}));

afterEach(() => {
  cleanup();
  vi.clearAllMocks();
});

describe('ChatCodeComposer', () => {
  it('renders composer with toolbar and textarea', () => {
    const { container } = render(
      <ChatCodeComposer
        value="Hola agentes"
        appendToDraft={vi.fn()}
        setDraft={vi.fn()}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('wraps selected text with code fences using shortcut', () => {
    const setDraft = vi.fn();
    const appendToDraft = vi.fn();
    render(
      <ChatCodeComposer
        value="Seleccion"
        appendToDraft={appendToDraft}
        setDraft={setDraft}
      />,
    );
    const textarea = screen.getByRole('textbox') as HTMLTextAreaElement;
    textarea.setSelectionRange(0, textarea.value.length);
    fireEvent.keyDown(textarea, { key: 'c', ctrlKey: true, shiftKey: true });
    expect(setDraft).toHaveBeenCalledWith('```\nSeleccion\n```\n');
    expect(appendToDraft).not.toHaveBeenCalled();
  });
});
