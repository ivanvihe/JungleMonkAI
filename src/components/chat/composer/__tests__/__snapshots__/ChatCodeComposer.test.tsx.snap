// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatCodeComposer > renders composer with toolbar and textarea 1`] = `
<div>
  <div
    class="chat-code-composer"
  >
    <div
      aria-label="Acciones del compositor"
      class="chat-code-toolbar"
      role="toolbar"
    >
      <button
        aria-label="Bloque de código"
        class="composer-toolbar-button"
        title="Bloque de código (Ctrl+Shift+\`)"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          🧱
        </span>
      </button>
      <button
        aria-label="Envolver selección"
        class="composer-toolbar-button"
        title="Envolver selección (Ctrl+Shift+C)"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          🪄
        </span>
      </button>
      <button
        aria-label="Copiar todo"
        class="composer-toolbar-button"
        title="Copiar todo"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          📋
        </span>
      </button>
      <div
        class="chat-code-toolbar-spacer"
      />
      <span
        aria-live="polite"
        class="chat-code-token-count"
      >
        3
         tokens estimados
      </span>
    </div>
    <textarea
      class="chat-code-textarea"
      rows="4"
    >
      Hola agentes
    </textarea>
  </div>
</div>
`;
